cart/domain/cart/paymentselection_test.go

```diff
@@ -41,6 +42,106 @@ func TestPrice_MarshalBinaryForGob(t *testing.T) {
+		name    string
+		split   cart.PaymentSplit
+		want    []byte
+		wantErr bool
+	}{
+		{
+			name: "marshall payment split",
+			split: func() cart.PaymentSplit {
+				result := cart.PaymentSplit{}
+				charge := domain.Charge{
+					Type: "t1",
+				}
+				firstQualifier := cart.SplitQualifier{
+					Method:     "m1",
+					ChargeType: charge.Type,
+				}
+				secondQualifier := cart.SplitQualifier{
+					Method:     "m2",
+					ChargeType: charge.Type,
+				}
+				result[firstQualifier] = charge
+				result[secondQualifier] = charge
+				return result
+			}(),
+			want:    []byte("{\"m1___t1\":{\"Price\":{\"Amount\":\"0\",\"Currency\":\"\"},\"Value\":{\"Amount\":\"0\",\"Currency\":\"\"},\"Type\":\"t1\"},\"m2___t1\":{\"Price\":{\"Amount\":\"0\",\"Currency\":\"\"},\"Value\":{\"Amount\":\"0\",\"Currency\":\"\"},\"Type\":\"t1\"}}"),
+			wantErr: false,
+		},
+		{
+			name: "marshall payment split with empty values - error",
+			split: func() cart.PaymentSplit {
+				result := cart.PaymentSplit{}
+				charge := domain.Charge{
+					Type: "t1",
+				}
+				firstQualifier := cart.SplitQualifier{
+					Method:     "m1",
+					ChargeType: "",
+				}
+				secondQualifier := cart.SplitQualifier{
+					Method:     "",
+					ChargeType: charge.Type,
+				}
+				result[firstQualifier] = charge
+				result[secondQualifier] = charge
+				return result
+			}(),
+			want:    nil,
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := tt.split.MarshalJSON()
+			if (err != nil) != tt.wantErr {
+				t.Errorf("PaymentSplit.MarshalJSON() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("PaymentSplit.MarshalJSON() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func TestPaymentSplit_UnmarshalJSON(t *testing.T) {
+	type args struct {
+		data []byte
+	}
+	tests := []struct {
+		name    string
+		args    args
+		wantErr bool
+	}{
+		{
+			name: "unmarshall payment split",
+			args: args{
+				data: []byte("{\"m1___t1\":{\"Price\":{\"Amount\":\"0\",\"Currency\":\"\"},\"Value\":{\"Amount\":\"0\",\"Currency\":\"\"},\"Type\":\"t1\"},\"m2___t1\":{\"Price\":{\"Amount\":\"0\",\"Currency\":\"\"},\"Value\":{\"Amount\":\"0\",\"Currency\":\"\"},\"Type\":\"t1\"}}"),
+			},
+			wantErr: false,
+		},
+		{
+			name: "unmarshall payment split empty method or type - error",
+			args: args{
+				data: []byte("{\"m1_t1\":{\"Price\":{\"Amount\":\"0\",\"Currency\":\"\"},\"Value\":{\"Amount\":\"0\",\"Currency\":\"\"},\"Type\":\"t1\"},\"m2___t1\":{\"Price\":{\"Amount\":\"0\",\"Currency\":\"\"},\"Value\":{\"Amount\":\"0\",\"Currency\":\"\"},\"Type\":\"t1\"}}"),
+			},
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			split := cart.PaymentSplit{}
+			if err := split.UnmarshalJSON(tt.args.data); (err != nil) != tt.wantErr {
+				t.Errorf("PaymentSplit.UnmarshalJSON() error = %v, wantErr %v", err, tt.wantErr)
+			}
```

Comment:
I would add an additional check to verify that the content of the unmarshalled split is actually correct.

Reply:
Good point, enahnced test